pkg/server/plugin/datastore/sql/sql.go

```diff
@@ -975,59 +888,127 @@ func fetchRegistrationEntry(tx *gorm.DB,
 	}, nil
 }
 
-func fetchRegistrationEntries(tx *gorm.DB,
-	req *common.Empty) (*datastore.FetchRegistrationEntriesResponse, error) {
+func listRegistrationEntries(tx *gorm.DB,
+	req *datastore.ListRegistrationEntriesRequest) (*datastore.ListRegistrationEntriesResponse, error) {
 
-	var entries []RegisteredEntry
-	if err := tx.Find(&entries).Error; err != nil {
-		return nil, sqlError.Wrap(err)
+	// list of selector sets to match against
+	var selectorsList [][]*common.Selector
+	if req.BySelectors != nil && len(req.BySelectors.Selectors) > 0 {
+		selectorSet := selector.NewSetFromRaw(req.BySelectors.Selectors)
+		if req.BySelectors.AllowAnyCombination {
+			// find entries matching any combination of the filter set
+			for combination := range selectorSet.Power() {
+				selectorsList = append(selectorsList, combination.Raw())
+			}
+		} else {
+			// find entries matching EXACTLY the filter set
+			selectorsList = append(selectorsList, selectorSet.Raw())
+		}
 	}
 
-	var sel []Selector
-	if err := tx.Find(&sel).Error; err != nil {
-		return nil, sqlError.Wrap(err)
+	// filter registration entries
+	entryTx := tx
+	if req.ByParentId != nil {
+		entryTx = entryTx.Where("parent_id = ?", req.ByParentId.Value)
+	}
+	if req.BySpiffeId != nil {
+		entryTx = entryTx.Where("spiffe_id = ?", req.BySpiffeId.Value)
 	}
 
-	// Organize the selectors for easier access
-	selectors := map[uint][]Selector{}
-	for _, s := range sel {
-		selectors[s.RegisteredEntryID] = append(selectors[s.RegisteredEntryID], s)
+	if len(selectorsList) == 0 {
+		// no selectors to filter against.
+		var entries []RegisteredEntry
+		if err := entryTx.Find(&entries).Error; err != nil {
+			return nil, sqlError.Wrap(err)
+		}
+
+		respEntries, err := modelsToEntries(tx, entries)
+		if err != nil {
+			return nil, sqlError.Wrap(err)
+		}
+
+		return &datastore.ListRegistrationEntriesResponse{
+			Entries: respEntries,
+		}, nil
 	}
 
-	// Populate registration entries with their related selectors
-	for _, entry := range entries {
-		if s, ok := selectors[entry.ID]; ok {
-			entry.Selectors = s
+	modelsSet := make(map[uint]RegisteredEntry)
+	for _, selectors := range selectorsList {
+		refCount := make(map[uint]int)
+		for _, s := range selectors {
+			var results []Selector
+			if err := tx.Find(&results, "type = ? AND value = ?", s.Type, s.Value).Error; err != nil {
+				return nil, sqlError.Wrap(err)
+			}
+
+			for _, r := range results {
+				if count, ok := refCount[r.RegisteredEntryID]; ok {
+					refCount[r.RegisteredEntryID] = count + 1
+				} else {
+					refCount[r.RegisteredEntryID] = 1
```

Comment:
Perhaps something like `refCount[r.RegisteredEntryID] = refCount[r.RegisteredEntryID] + 1` to avoid an if/else?

Reply:
sure. will do.